bash_completion

```diff
@@ -2307,33 +2307,83 @@ _comp_command_offset()
+
+                    # restart completion (once) if function exited with 124
+                    if (($? == 124 && retry_count++ == 0)); then
+                        # Note: When the completion function returns 124, the
+                        # state of COMPREPLY is discarded.
+                        COMPREPLY=()
+
+                        cspec=$(complete -p "$compcmd" 2>/dev/null)
+
+                        # Note: When completion spec is removed after 124, we
+                        # do not generate any completions including the default
+                        # ones. This is the behavior of the original Bash
+                        # progcomp.
+                        [[ $cspec ]] || break
+
+                        continue
+                    fi
                 else
-                    $func "$cmd" "${COMP_WORDS[-1]}"
-                fi
+                    # complete -C <command>
 
-                # restart completion (once) if function exited with 124
-                if (($? == 124 && retry_count++ == 0)); then
-                    # Note: When the completion function returns 124, the state
-                    # of COMPREPLY is discarded.
-                    COMPREPLY=()
+                    # get command name
+                    local completer=${cspec#* -C \'}
 
-                    cspec=$(complete -p "$compcmd" 2>/dev/null)
+                    # completer commands are always single-quoted
+                    if ! _comp_dequote "'$completer"; then
+                        _minimal
+                        break
+                    fi
+                    completer=${ret[0]}
+
+                    local -a suggestions
+
+                    local reset_monitor=$(shopt -po monitor) reset_lastpipe=$(shopt -p lastpipe)
+                    set +o monitor
+                    shopt -s lastpipe
+
+                    (
+                        export COMP_KEY COMP_LINE COMP_POINT COMP_TYPE
 
-                    # Note: When completion spec is removed after 124, we do
-                    # not generate any completions including the default ones.
-                    # This is the behavior of the original Bash progcomp.
-                    [[ $cspec ]] || break
+                        if ((${#COMP_WORDS[@]} >= 2)); then
+                            $completer "$cmd" "${COMP_WORDS[-1]}" \
+                                "${COMP_WORDS[-2]}"
+                        else
+                            $completer "$cmd" "${COMP_WORDS[-1]}"
+                        fi
+                    ) | mapfile suggestions
 
-                    continue
+                    $reset_monitor
+                    $reset_lastpipe
+
+                    local suggestion
+                    local i=0
+                    COMPREPLY=()
+                    for suggestion in "${suggestions[@]}"; do
+                        # `mapfile -t` was added in bash 4.4, so we have to
+                        # strip trailing newlines manually
```

Comment:
Are we sure about this? We've used `mapfile -t` in the `cvs` completion since 2011 (e35eabb791733996078aae9161fc01ab5f94a9e3), I think 5 years before bash 4.4 was released.

Reply:
The `NEWS` file indicates that [it was added in 4.4][news-4.4], but doing some digging, that's it looks like that notice was misleading. Looking at the source code, it looks like the flag did exist starting with bash 4.0, but originally [only removed trailing newlines][mapfile-4.0]; bash 4.4 added the `-d` flag and updated `-t` to remove the specified delimiter when both flags are used. I'll update this accordingly.

[news-4.4]: https://github.com/bminor/bash/blob/bash-4.4/NEWS#L17-L19
[mapfile-4.0]: https://github.com/bminor/bash/blame/bash-4.0/builtins/mapfile.def#L37